src/iso.c

```diff
@@ -0,0 +1,494 @@
+
+        // Don't actually want to read off the end of the file
+        bytes_to_file_read = file->data_length - file->cur_offset + 1;
+        bytes_to_disk_read = bytes_to_file_read - 1;
+        file->at_EOF = 1;
+    }
+    int bytes_read = iso_media_read(dest, 1, bytes_to_disk_read, iso_p);
+    if (bytes_read != bytes_to_disk_read) {
+        console_printf("file reading error\n");
+        iso_media_close(iso_p);
+        return -1;
+    }
+
+    iso_media_close(iso_p);
+    file->cur_offset += bytes_to_file_read;
+    return bytes_read;
+}
+
+/**
+ * @brief Acts as an fclose() for iso disk
+ * @details Frees the iso_point that was used to navigate the ISO disk.
+ *
+ * @param iso_p A pointer to the iso_point to close.
+ * @return Pointer to an allocated iso_point with members initialized to 0
+ */
+static void iso_media_close(struct iso_point *iso_p) {
+    kfree(iso_p);
+    return;
+}
+
+/**
+ * @brief Treating the iso on the given ata_unit as a file, opens it
+ * @details kmalloc's and returns an iso_point which is set to the first
+ * byte of the given ata_unit
+ *
+ * @param ata_unit The ata unit to open the iso_point on
+ * @return A pointer to the iso_point at the first byte of the iso.
+ */
+static struct iso_point *iso_media_open(int ata_unit) {
+    struct iso_point *to_return = kmalloc(sizeof(struct iso_point));
+    to_return->cur_extent = 0;
+    to_return->cur_offset = 0;
+    to_return->ata_unit = ata_unit;
+    return to_return;
+}
+
+/**
+ * @brief Reads in the specified amount from the iso_p
+ * @details Reads into dest the specified amount from the iso starting
+ * at the current extent and offset of iso_p
+ *
+ * @param dest A buffer to put the read information into
+ * @param elem_size The size in bytes of each element to be read
+ * @param num_elem The number of elements to read into dest
+ * @param iso_p Pointer to the struct iso_point to begin reading at.
+ * @return The number of elements read in, -1 on error.
+ */
+static int iso_media_read(void *dest, int elem_size, int num_elem, struct iso_point *stream) {
+    int bytes_needed = num_elem * elem_size;
+    if (bytes_needed <= 0) {
+        return 0;
+    }
+
+    int to_read_from_start_of_cur_extent = stream->cur_offset + bytes_needed;
+    int atapi_blocks_to_read = to_read_from_start_of_cur_extent / ATAPI_BLOCKSIZE;
+
+    if (to_read_from_start_of_cur_extent % ATAPI_BLOCKSIZE) {
+        //integer division rounds down, but we need to round up
+        //if there is rounding to be done
+        atapi_blocks_to_read++;
+    }
+
+    //ensure that we need to perform the read before we do
+    if (atapi_blocks_to_read > 1 ||
+        global_atapi_extent != stream->cur_extent ||
+        global_atapi_unit != stream->ata_unit) {
+        char buffer[atapi_blocks_to_read * ATAPI_BLOCKSIZE];
+        if (!atapi_read(stream->ata_unit, buffer, atapi_blocks_to_read, stream->cur_extent)) {
+            console_printf("atapi_read is 0 in iso_media_read()\n");
```

Comment:
This seems more like a debugging line than a useful error message to the user. Correct me if I'm wrong tho!